### ISSUE TITLE
How do I control the amount of memory used by the program

### ISSUE THREAD

**Issue by qazwsx1995** [voichek/kmersGWAS#48]
How do I control the amount of memory used by the program  with the script “kmers_add_strand_information” ，it used about 200G each samples ，how to change the script or add script parameters,thank you~

**Comment by qazwsx1995** [comment]
@voichek

**Comment by voichek** [comment]
It sounds like way way too much, for ~1000 genomes of A. thaliana (and ~500M k-mers) it took 1.6GB, how many individuals do you have? and with what parameters you run the pipeline?

**Comment by voichek** [comment]
Sorry, you meant kmers_add_strand, this also sound like a lot, how many reads do you have in that specific sample?

**Comment by qazwsx1995** [comment]
"$library_dir/bin/kmers_add_strand_information -c $kmc_canon -n $kmc_non_canon -k $K -o $sample_dir/kmers_with_strand > $sample_dir/add_strand.log.out" 
 Every time I run an individual, samples_1.fastq is about 70G, which is an individual pig.

**Comment by qazwsx1995** [comment]
about 200 million reads @voichek

**Comment by voichek** [comment]
There is a way to decrease the the memory size but it requires changing the code slightly and compiling
The program now runs with KmerUint64Hash and if it is changed to KmerUint64SparseHash it will use much less memory (in the cost of a reduction in speed in this step).

Another question, what is your coverage? and what is your threshold on the number of appearances of a k-mer need to appear when you run KMC (canon)?

**Comment by qazwsx1995** [comment]
THRESHOLD_COUNT=2 #threshold to count k-mers per sample

**Comment by voichek** [comment]
Coverage? If you have high coverage maybe you should take the threshold a bit up.

But for the general question, in this step if you want to lower the memory you need to change the code. All the next steps use much less memory, this is the only point in the process where I upload all the k-mers to the memory.

**Comment by qazwsx1995** [comment]
I will look for coverage again. There is one last question. What standard should the relationship between coverage and threshold have? Could  you give an empirical function or a correlation graph?

**Comment by voichek** [comment]
The quick answer is that I have not checked this systematically. 
Practically, in our work we took a threshold of 3 for _A. thaliana_ where coverage was high, and lowest accessions had around 10x. For maize and tomato we used a threshold of 2 as coverage of lowest was around 6x. This was chosen by intuition and not systematically. I would say that if coverage is not extremely high, a threshold of 2 or 3 should work well as it is unlikely that the same mutation will create the same k-mer multiple times and in different strains.

Hope it helps.
